### DISCUSSION TITLE
How creating crafting table result slot event.

### DISCUSSION THREAD

**Discussion by Intezium** [FabricMC/fabric-api#2396] [Mod Dev Support] [answered]
I'm trying to create an event before giving an item to a player in the crafting table.
Everything looks fine, there are no errors. But the event is not triggered.
I also tried these options `target` by this tutorial [Link](https://jenkins.liteloader.com/job/Mixin/javadoc/index.html?org/spongepowered/asm/mixin/injection/struct/MemberInfo.html):
`"Lnet/minecraft/entity/player/PlayerEntity;*(Lnet/minecraft/item/ItemStack;Z)Lnet/minecraft/entity/ItemEntity;"`
`"Lnet/minecraft/entity/player/PlayerEntity;dropItem*"`

`CraftingResultSlotMixin.java`
```
@Mixin(CraftingResultSlot.class)
public class CraftingResultSlotMixin {
    @Inject(at = @At(
            value = "INVOKE",
            target = "Lnet/minecraft/entity/player/PlayerEntity;dropItem(Lnet/minecraft/item/ItemStack;Z)Lnet/minecraft/entity/ItemEntity;"),
            method = "onTakeItem(Lnet/minecraft/entity/player/PlayerEntity;Lnet/minecraft/item/ItemStack;)V",
            cancellable = true)
    private void onTakeItem(PlayerEntity player, ItemStack itemStack, CallbackInfo info) {
        ActionResult result = TakeCraftingItemCallback.EVENT.invoker().interact(player, itemStack);

        if (result == ActionResult.FAIL) {
            info.cancel();
        }
    }
}
```
`TakeCraftingItemCallback.java`
```
public interface TakeCraftingItemCallback {
    Event<TakeCraftingItemCallback> EVENT = EventFactory.createArrayBacked(TakeCraftingItemCallback.class,
            (listeners) -> (player, stack) -> {
                for (TakeCraftingItemCallback listener : listeners) {
                    ActionResult result = listener.interact(player, stack);

                    if(result != ActionResult.PASS) {
                        return result;
                    }
                }

                return ActionResult.PASS;
            });

    ActionResult interact(PlayerEntity player, ItemStack stack);
}
```
`TutorialMod.java`
```
public class TutorialMod implements ModInitializer {
    @Override
    public void onInitialize() {
        TakeCraftingItemCallback.EVENT.register((player, stack) ->
        {
            player.sendMessage(Text.translatable("crafting.event"));

            return ActionResult.PASS;
        });
    }
}
```
`tutorial.mixins.json`
```
{
  "required": true,
  "minVersion": "0.8",
  "package": "net.fabricmc.example.mixin",
  "compatibilityLevel": "JAVA_17",
  "mixins": [
    "CraftingResultSlotMixin"
  ],
  "client": [],
  "server": [],
  "injectors": {
    "defaultRequire": 1
  }
}
```

**Comment by warjort** [comment, marked as answer]
Please don't post minecraft source here. Just reference the method name and describe the part you are interested in.

The dropItem() is for when the item doesn't fit in the player's inventory, it drops the item on the ground.

The onCrafted() is to allow items to add default nbt for newly crafted items and to update crafting statistics.
The loop is is to update the crafting grid to decrease/remove items used in the craft.

If you want to detect the player taking an item from the result slot just @Inject at the HEAD of the onTakeItem() method.
i.e. the start of that method before all the processing of the craft starts
